hdk-proc-macros/src/into_zome.rs

```diff
@@ -0,0 +1,503 @@
+                                && !is_tagged_with(RECEIVE_CALLBACK_ATTRIBUTE)(func)
+                        } else {
+                            true // and anything that is not a function
+                        }
+                    })
+                    .collect()
+            }
+            None => Vec::new(),
+        }
+    }
+
+    fn extract_receive_callback(&self) -> Option<ReceiveCallback> {
+        // find all the functions tagged as the genesis callback
+        let callbacks: Vec<ReceiveCallback> = funcs_iter(self)
+            .filter(is_tagged_with(RECEIVE_CALLBACK_ATTRIBUTE))
+            .fold(Vec::new(), |mut acc, func| {
+                let inputs = func.decl.inputs;
+
+                match inputs.len() {
+                    1 => {
+                        let input = inputs.iter().next().unwrap();
+                        if let syn::FnArg::Captured(arg) = input {
+                            let name = match &arg.pat {
+                                syn::Pat::Ident(name_ident) => name_ident.ident.clone(),
+                                _ => {
+                                    emit_error(
+                                        &func.ident,
+                                        "The argument to receive must have a name",
+                                    );
+                                    panic!()
+                                }
+                            };
+                            acc.push(ReceiveCallback {
+                                param: name,
+                                code: *func.block,
+                            });
+                        }
+                    }
+                    _ => {
+                        emit_error(
+                            &func.ident,
+                            "Receive callback must take a single argument of type 'String'",
+                        );
+                        panic!()
+                    }
+                }
+                acc
+            });
+        match callbacks.len() {
+            0 => None,
+            1 => Some(callbacks[0].clone()),
+            _ => {
+                emit_error(
+                    &self.ident,
+                    "Multiple functions tagged with receive. Only one permitted per zome.",
+                );
+                panic!()
+            }
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use syn::parse_quote;
+
+    #[test]
+    fn test_extract_genesis_smoke_test() {
+        let module: syn::ItemMod = parse_quote! {
+            mod zome {
+                #[genesis]
+                fn genisis() {
+                    Ok(())
+                }
+            }
+        };
+        let _ = module.extract_zome();
+    }
+
+    #[test]
+    fn test_extract_single_trait() {
+        let module: syn::ItemMod = parse_quote! {
+            mod zome {
+                #[genesis]
+                fn genisis() {
```

Comment:
Same thing: is this intentionally genisis not genesis?

Reply:
> @maackle
> 
> hm... if the two really do match, maybe we could have one set of js tests and test both against that...
> we were thinking of spending a reasonably long period of time just refining the proc macro hdk before announcing it, fully documenting it, etc, so there will be a period where they both stay in the repo

That is how it is set up now! The tests are copied from `app_spec` to `app_spec_proc_macro` when the tests run because of the limitation of `hc test` that they must be in the same dir.